File: Go/Day24.go
package main

import AH "./adventhelper"

type Pos struct {
	X, Y int
}

func (pt Pos) nbrs() (ns []Pos) {
	ns = []Pos{Pos{X:pt.X + 1, Y:pt.Y    },
	           Pos{X:pt.X,     Y:pt.Y + 1},
	           Pos{X:pt.X - 1, Y:pt.Y    },
	           Pos{X:pt.X,     Y:pt.Y - 1},
	           Pos{X:pt.X - 1, Y:pt.Y + 1},
	           Pos{X:pt.X + 1, Y:pt.Y - 1}}

	return
}

func parseLine(x0 int, y0 int, s string) (Pos) {
	x1, y1 := x0, y0

	rs := []rune(s)

	for i := 0; i < len(rs); i++ {
		// grab first char
		h := rs[i]
		switch h {
			case rune('e') : x1++
			case rune('w') : x1--
			case rune('n') : {
				i++
				hh := rs[i]
				switch hh {
					case rune('e') : y1++
					case rune('w') : x1--; y1++		
				}	
			}
			case rune('s') : {
				i++
				hh := rs[i]
				switch hh {
					case rune('e') : x1++; y1--
					case rune('w') : y1--
				}
			}
		}
	}

	return Pos{X:x1, Y:y1}
}

func (pt Pos) nbrsCount(space map[Pos]int) (count int) {
	count = 0
	ns := pt.nbrs()
	for _, n := range ns {
		if _, ok := space[n]; ok {
			count++
		}
	}
	return
}

func tick (space map[Pos]int) (newSpace map[Pos]int) {
	newSpace = make(map[Pos]int)

	for pos, _ := range space {
		// check if this cell surives
		nCount := pos.nbrsCount(space) // the number of neighbours
		if nCount == 1 || nCount == 2 {
			newSpace[pos] = 1
		}
		// now check its neighbours
		neighbours := pos.nbrs()
		for _, nPos := range neighbours {
			nCount = nPos.nbrsCount(space) // the number of neighbours
			// if we've already done this point...
			if _, ok := newSpace[nPos]; ok {
				// do nothing..
			} else { // we haven't checked this point yet
				if _, ok := space[nPos]; ok { // Is it currently alive..
					if nCount == 1 || nCount == 2 { //..and has 2 or 3 live neighbours...
						newSpace[nPos] = 1 // ..send it to next generation
					} // otherwise it dies here
				} else { // It is NOT currently alive...
					if nCount == 2 { //..and has exactly 3 live neighbours...
						newSpace[nPos] = 1 // ..send it to next generation
					} // otherwise it dies here
				}
			}
		}
	}

	return
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input24.txt")

	points := make(map[Pos]int)

	for _, s := range ss {
		pt := parseLine(0,0,s)
		if _, ok := points[pt]; ok {
			delete(points, pt)
		} else {
			points[pt] = 1
		}
	}
	part1 := len(points)

	for i := 0; i < 100; i++ {
		points = tick(points)
	}	
	part2 := len(points)

	AH.PrintSoln(24, part1, part2)

	return
}

File: C++/Day24.cpp
#include "AH.h"

namespace Day24
{

	class Pos
	{
	public:
		Pos(const int64_t x, const int64_t y) : X(x), Y(y) {};
		int64_t X;
		int64_t Y;

		std::vector<Pos> nbrs() const;
		uint64_t nbrsCount(const std::set<Pos>& space) const;
		bool operator<(const Pos& rhs) const;
	};

	bool Pos::operator<(const Pos& rhs) const
	{
		if (X != rhs.X)
			return (X < rhs.X);

		if (Y != rhs.Y)
			return (Y < rhs.Y);

		return false;
	}

	std::vector<Pos> Pos::nbrs() const
	{
		std::vector<Pos> ns{Pos(X+1,Y), Pos(X,Y+1),Pos(X-1,Y),Pos(X,Y-1),
		                    Pos(X-1,Y+1),Pos(X+1,Y-1)};

		return ns;
	}

	Pos parseLine(int64_t x, int64_t y, const std::string& s)
	{
		for (size_t i = 0; i < s.size(); ++i)
		{
			auto h = s.at(i);
			if (h == 'e') { x++; }
			else if (h == 'w') { x--; }
			else if (h == 'n')
			{
				i++;
				auto hh = s.at(i);
				if (hh == 'e') { y++; }
				else if (hh == 'w') { x--; y++; }
			}
			else if (h == 's')
			{
				i++;
				auto hh = s.at(i);		
				if (hh == 'e') { x++; y--; }
				else if (hh == 'w') { y--; }		
			}
		}

		return Pos(x, y);  
	}

	uint64_t Pos::nbrsCount(const std::set<Pos>& space) const
	{
		uint64_t count = 0;
		auto ns = nbrs();

		for ( auto n : ns)
			if (space.find(n) != space.end())
				count++;		

		return count;
	}

	void tick(std::set<Pos>& space)
	{
		std::set<Pos> newSpace;
		for ( auto pt : space )
		{
			auto nCount = pt.nbrsCount(space);
			if ((nCount == 1) || (nCount == 2))
				newSpace.insert(pt);

			auto ns = pt.nbrs();
			for ( auto nPos : ns)
			{
				nCount = nPos.nbrsCount(space);
				if (newSpace.find(nPos) != newSpace.end())
				{
					// do nothing
				}
				else
				{
					if (space.find(nPos) != space.end())
					{
						if ((nCount == 1) || (nCount == 2))
							newSpace.insert(nPos);
					}
					else
					{
						if (nCount == 2)
							newSpace.insert(nPos);
					}
				}
			}
		}

		space = newSpace;
		return;
	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);
		std::set<Pos> space;

		for (auto s: inputLines)
		{
			auto pt = parseLine(0,0,s);
			if (space.find(pt) != space.end())
				space.erase(pt);
			else
				space.insert(pt);
		}
		auto part1 = space.size();

		for (size_t i = 0; i < 100; ++i)
			tick(space);

		auto part2 = space.size();;

		AH::PrintSoln(24, part1, part2);

		return 0;
	}

}
